Added pade approximant functionality

LylePoley opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #27665

<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Added funcitonality for computing pade approximants to expressions, which has been discussed in #27665, #27661, and #27115. For a function $f$ and integers $m$ and $n$, the $[m/n]$ pade approximant of $f$ is the rational function $p(x)/q(x)$, where $p(x)$ and $q(x)$ are polynomials of degree at most $m$ and $n$ respectively. The pade approximation, when it exists, is such that the taylor series of $p(x)/q(x)$ around $x=0$ matches the taylor series of $f(x)$ up to at least order $(m + n)$ in $x$. For example, we can show that 

$$
\frac{x^2 + 6x + 12}{x^2 - 6x + 12} = 1 + x + \frac{1}{2}x^2 + \frac{1}{6}x^3 + \frac{1}{24}x^4 + O\left(x^5\right).
$$

Therefore, noting that the right hand side is the Taylor expansion of $e^x$ up to order $4$, we can conclude that the left hand side of the above equation is equal to the $[2/2]$ pade approximant for $e^x$. To demonstrate this in sympy, we can now use the function `pade_approximant`:

 ```python 
from sympy.series.approximants import pade_approximant
from sympy.abc import x
from sympy import exp

print(pade_approximant(exp(x), x, x0=0, m=2, n=2))
```
```
(x**2 + 6*x + 12)/(x**2 - 6*x + 12)
```

To facilitate this, the following functions have also been added:
- `sympy.polys.polytools.gcdex_steps`
- `sympy.polys.polytools._gcdex_steps_domain`
- `sympy.polys.polytools._gcdex_steps_polynomial`
- `sympy.series.approximants.pade_approximant_gcdex`
- `sympy.series.approximants.pade_approximants_gcdex`

`gcdex_steps` performs the extended euclidean algorithm on two polynomial inputs, yielding the results of the intermediate calculations. `_gcdex_steps_domain` and `_gcdex_steps_polynomial` are helper functions. For example:

```python 
from sympy.polys.polytools import gcdex_steps

f = x**4 - 2*x**3 - 6*x**2 + 12*x + 15
g = x**3 + x**2 - 4*x - 4

for step in gcdex_steps(f, g): 
    print(step)
```
```
(0, 1, x**3 + x**2 - 4*x - 4)
(1, 3 - x, x**2 + 4*x + 3)
(3 - x, x**2 - 6*x + 10, 5*x + 5)
```

`pade_approximants_gcdex` computes all pade approximants of a given order. This function returns all pade approximants as pairs of polynomials
```python
from sympy.series.approximants import pade_approximants_gcdex

exp_taylor_poly = Poly(1 + x + x**2/2 + x**3/6 + x**4/24, x)
pade_exp = pade_approximants_gcdex(exp_taylor_poly, 2)
for p in pade_exp: print(p)
```
```
(Poly(1/2*x**2 + x + 1, x, domain='QQ'), Poly(1, x, domain='QQ'))
(Poly(2*x + 4, x, domain='QQ'), Poly(-2*x + 4, x, domain='QQ'))
(Poly(1, x, domain='QQ'), Poly(1/2*x**2 - x + 1, x, domain='QQ'))
```

the above are the `[2/0], [1/1]` and `[0/2]` pade approximants to $e^x$. The singular `pade_approximant_gcdex` calls `pade_approximants_gcdex` and exits when the requested approximant is computed.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* series
  * Added functions `pade_approximant`, `pade_approximants_gcdex`, and `pade_approximant_gcdex`. `pade_approximant(f, x, x0, m, n)` computes the `[m/n]` pade approximant to the expression `f(x)` around the point `x0`.  `pade_approximant_gcdex` does the same for polynomials and `pade_approximants_gcdex` computes all pade approximants to a certain order for a given polynomial. `pade_approximants_gcdex` is used internally by `pade_approximant_gcdex`, and `pade_approximant_gcdex` is used internally by `pade_approximant`.

* polys
  * Added functions `gcdex_steps`, `_gcdex_steps_domain`, and `_gcdex_steps_polynomial`. `gcdex_steps(p, q)` returns all intermediate steps of the extended euclidean algorithm for the given `p` and `q`. If $p$ and `q` are convertible to polynomials, then `_gcdex_steps_polynomial` is called and if they are not convertible to polynomials (e.g. if `p` and `q` are integers) then the euclidean algorithm is attempted with`_gcdex_steps_domain`. `gcdex_steps` is used internally by `pade_approximants_gcdex`.

<!-- END RELEASE NOTES -->

LylePoley committed:
series: Added a function pade_approximant and tests

The pade approximant is similar to a taylor series, although rather than approximating an expression with a polynomial of given degree, a pade approximant approximates an expression with a rational function whose numerator and denominator each have given degree.

Added a pade_approximant function. This function allows users to easily compute the [n/m] pade approximant of a given expression with a similar interface to computing a series expansion for an expression. I wrote this because I used to use the function "PadeApproximant" frequently in mathematica and found it very useful. The code for computing the approximant is adapted from the source code for scipy.interpolate.pade.

Example:

In [1]: import sympy as sp

In [2]: x = sp.symbols('x')

In [3]: pade_approximant(sp.exp(x), x, 0, 2, 1)

Out [3]: (x**2/6 + 2*x/3 + 1)/(1 - x/3)

The command "pade_approximant(sp.exp(x), x, 0, 2, 1)" computes the [2/1] approximant of exp(x) around the point x = 0.

LylePoley committed:
Added three functions for computing pade approximants (issue #27665)
    1. sympy.polys.polytools.extended_euclidean_algorithm(f, p)

    which takes a pair of polynomials (f, p) and generates the succesive steps
    of the extended euclidean algorithm (s, t, r). This has been added as the
    steps, rather than just the final result, are needed to compute pade
    approximants.

    2. sympy.series.approximants.pade_approximants(f, x, order)

    This function generates all pade approximants of the given order to the
    function. pade_approximants(f, x, order) relies on function 1.

    3. sympy.series.approximants.pade_approximant(f, x, m, n)

    Computes the [m/n] pade approximant to the function f(x) if it exists.
    If there is no [m/n] pade approximant, the function returns None.

Examples:
    In [1]: from sympy import sin, Poly
    In [2]: from sympy.abc import x

    In [3]: from sympy.polys.polytools import extended_euclidean_algorithm
    In [4]: from sympy.series.approximants import pade_approximant, pade_approximants

    In [5]: f = Poly((x + 1)*(x + 2), x)
    In [6]: g = Poly((x + 1)*(x + 3), x)
    In [7]: eea_result = extended_euclidean_algorithm(f, g)
    In [8]: for s, t, r in eea_result: print(s, t, r)
    Out [8]: Poly(1, x, domain='ZZ') Poly(-1, x, domain='ZZ') Poly(-x - 1, x, domain='ZZ')

    In [9]: pade_sin = pade_approximants(sin(x), x, 5)
    In [10]: for num, denom in pade_sin: print(num/denom)
    Out [10]: x**5/120 - x**3/6 + x
              -(20*x**4 - 120*x**2)/(120*x)
              (-7*x**3/60 + x)/(x**2/20 + 1)
              360*x**2/(7*(60*x**3/7 + 360*x/7))
              x/(7*x**4/360 + x**2/6 + 1)

    In [11]: pade_exp = pade_approximant(sin(x), x, 3, 2)
    In [12]: pade_exp
    Out [12]: (-7*x**3/60 + x)/(x**2/20 + 1)

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -5307,6 +5307,105 @@ def half_gcdex(f, g, *gens, **args):
 
 
+@public
+def extended_euclidean_algorithm(f, g):
```
In SymPy the extended Euclidean algorithm is referred to as `gcdex`. There are a few functions with this name including one in this file.

Probably this function should be called something like `igcdex` since it returns an iterator over the polynomial remainder sequence.

oscarbenjamin replied:
Or perhaps `gcdex_steps` or something.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -5307,6 +5307,105 @@ def half_gcdex(f, g, *gens, **args):
+    Description
+    ===========
+
+    Implements the extended euclidean algorithm for polynomials as described in e.g.
+        McEliece, R. J., & Shearer, J. B. (1978).
+        A Property of Euclid's Algorithm and an
+        Application to Pade Approximation.
+        SIAM Journal on Applied Mathematics, 34(4), 611-615.
+        doi:10.1137/0134048
```
References should go in a section at the end of the docstring:
https://docs.sympy.org/latest/contributing/docstring.html

haru-44 commented on sympy/series/approximants.py:
```diff
@@ -101,3 +104,146 @@ def approximants(l, X=Symbol('x'), simplify=False):
+    x/(7*x**4/360 + x**2/6 + 1)
+    """
+
+    assert order >= 0, "order must be a non-negative integer"
```
When handling invalid user input, please use `raise`.

oscarbenjamin commented on sympy/series/approximants.py:
```diff
@@ -101,3 +104,146 @@ def approximants(l, X=Symbol('x'), simplify=False):
+@public
+def pade_approximants(
+    f: Expr, x: Expr, order: int
+) -> Generator[tuple[Poly, Poly], None, None]:
```
There needs to be `from __future__ import annotations` at the top of the file for this kind of annotation to work in older Python versions.

LylePoley commented:
Sorry to take so long. I believe that I have adressed the issues within the code itself that have been suggested.

LylePoley committed:
Fixed some issues raised in pull request sympy#27671

oscarbenjamin commented on sympy/series/approximants.py:
```diff
@@ -101,3 +106,159 @@ def approximants(l, X=Symbol('x'), simplify=False):
+    if n is None:
+        n = m
+
+    assert m >= 0 and n >= 0, "m and n must be non-negative integers"
```
This should use e.g. `raise ValueError`. User input should not be validate with assert.

oscarbenjamin replied:
Otherwise I think this looks good.

LylePoley committed:
Removed user validation with assert, replaced with raise (pull request sympy#27671)

oscarbenjamin commented on sympy/series/approximants.py:
```diff
@@ -101,3 +106,160 @@ def approximants(l, X=Symbol('x'), simplify=False):
+    ========
+
+    sympy.series.approximants.approximants
+    sympy.series.approximants.pade_approximant
```
We should add See Also in approximants as well.

It looks like approximants is imported at top-level like
```python
from sympy import approximants
```
I wonder if either of these functions would make sense to be imported at top-level.

LylePoley replied:
Agree to adding to the see also in approximants. I personally would find it useful to have `pade_approximant` imported at the top level, I'm not sure about `pade_approximants`. I don't know what the convention is for what goes in to the top level though, happy to defer to you on that. 

To make `pade_approximant` importable at the top level, am I correct in thinking that I have to add `pade_approximant` to the list of imports in `sympy.series.__init__.py` and also to the list of imports in `sympy.__init__.py`?

oscarbenjamin replied:
> I personally would find it useful to have `pade_approximant`

Let's add that one then.

> am I correct in thinking

Yes

LylePoley replied:
Done

LylePoley committed:
Added `pade_approximant` to top level imports

LylePoley committed:
Adding `pade_approximant` to __all__

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -5307,6 +5307,113 @@ def half_gcdex(f, g, *gens, **args):
+
+    Here, the final value of r is -(x + 1), so this is the gcd of f and g
+
+    See Also
+    ========
+
+    sympy.polys.polytools.gcdex: Extended Euclidean algorithm witout intermediate steps.
+    sympy.polys.polytools.half_gcdex: Half extended Euclidean algorithm.
```
Sorry to do this again but I think we also need See Also references from gcdex to here.

We need See Also reference in both directions everywhere. They are often more useful than anything else in the docs.

oscarbenjamin replied:
Otherwise I think this looks good.

LylePoley replied:
No problem, what do you think to also referencing `polys.euclidtools.dup_gcdex`?

LylePoley replied:
Adding the see also reference to `gcdex_steps` has led to a failed test that I'm not sure how to fix. Can you advise?

oscarbenjamin replied:
I think you need to add `gcdex_steps` to the docs:
https://github.com/sympy/sympy/blob/b9321466422048110da8c6d8822742c26df1f256/doc/src/modules/polys/reference.rst?plain=1#L34

Possibly the other functions here should be added to the docs as well.

LylePoley replied:
Ahh ok. I have added `gcdex_steps` to the docs, as well as all three functions from `sympy.series.approximants` (the one that was there before and the two I have written). I've put those in a new file `sympy/doc/src/modules/series/approximants.rst`.

LylePoley committed:
Added `see also` references and docstring formatting modifications.

LylePoley committed:
added `sympy.polys.polytools.gcdex_steps`

LylePoley committed:
Added functions in `sympy.series.approximants` to the docs.

oscarbenjamin commented:
Looks like there are problems with the existing docstrings now that they are added to the docs:
```
The docs build failed with the following errors:
/home/runner/work/sympy/sympy/doc/src/modules/series/approximants.rst: WARNING: document isn't included in any toctree [toc.not_included]
/home/runner/work/sympy/sympy/sympy/series/approximants.py:docstring of sympy.series.approximants.approximants:50: WARNING: py:obj reference target not found: sympy.concrete.guess.guess_generating_function_rational [ref.obj]
/home/runner/work/sympy/sympy/sympy/series/approximants.py:docstring of sympy.series.approximants.approximants:50: WARNING: py:obj reference target not found: mpmath.pade [ref.obj]
```
I think the new page needs to be listed from another page so that there is a link to it. It is possible that the guess_generating_function also needs to be added to the docs. The mpmath reference might be resolvable but it might need have a full path to the pade function.

LylePoley committed:
Added `approximants.rst`

LylePoley committed:
Added functions to docs

sympy.concrete.guess.guess_generating_function
sympy.concrete.guess.guess_generating_function_rational

LylePoley committed:
Added cross references

LylePoley commented:
> I think the new page needs to be listed from another page so that there is a link to it. It is possible that the guess_generating_function also needs to be added to the docs. The mpmath reference might be resolvable but it might need have a full path to the pade function.

I think that `guess_generating_function` does need to be added to the docs as you say, as do any functions that this function references and so on. 

There appear to be a lot of functions missing from the .rst files. Is this deliberate? If it isnt, is it an open issue?

LylePoley committed:
Attempt to fix cross references

LylePoley committed:
Attempt to fix cross-links

oscarbenjamin commented:
> There appear to be a lot of functions missing from the .rst files. Is this deliberate? If it isnt, is it an open issue?

It isn't deliberate. There may be issues about this for some particular parts of the docs but generally whenever I find that something is missing as you are right now I just add them to the docs rather than opening an issue.

It would be better if the docs could be generated fully automatically but autodoc is still only sort of semi-automatic.

oscarbenjamin commented on sympy/series/approximants.py:
```diff
@@ -53,7 +58,9 @@ def approximants(l, X=Symbol('x'), simplify=False):
 
     sympy.concrete.guess.guess_generating_function_rational
-    mpmath.pade
+    :py:obj:`!mpmath.pade`
```
I'm not sure what the syntax is for intersphinx references and I'm also not sure that mpmath.pade is the correct reference for the function.

I suggest just deleting this mpmath reference.

oscarbenjamin committed:
doc(concrete): add guess functions to the docs

oscarbenjamin committed:
fmt(concrete): reflow text in docstrings

oscarbenjamin commented:
Do the examples shown at https://github.com/sympy/sympy/issues/27665#issuecomment-2849177889 represent some issue with the implementation in this PR?

LylePoley commented:
I don't believe so. I've tested the first proposed failed example and the implementation in this PR works as it should. I've added a comment to that thread to this effect.

LylePoley committed:
refactoring existing code

oscarbenjamin commented on sympy/series/approximants.py:
```diff
@@ -101,3 +108,282 @@ def approximants(l, X=Symbol('x'), simplify=False):
+    if order < 0:
+        raise ValueError("'order' must be a non-negative integer")
+
+    # truncate f so f.degree() <= order
+    if f.degree() > order:
+        coefficients = f.all_coeffs()
+        f = Poly.from_list(coefficients[-(order + 1):], gens=x)
```
It isn't very discoverable but:
```
In [14]: p = Poly([1,1,1,1,1,1], x)

In [15]: p
Out[15]: Poly(x**5 + x**4 + x**3 + x**2 + x + 1, x, domain='ZZ')

In [16]: p.slice(0, 3)
Out[16]: Poly(x**2 + x + 1, x, domain='ZZ')
```

oscarbenjamin commented on sympy/series/approximants.py:
```diff
@@ -101,3 +108,282 @@ def approximants(l, X=Symbol('x'), simplify=False):
+    if m < 0 or n < 0:
+        raise ValueError("m and n must be non-negative integers")
+
+    if isinstance(f, Poly):
+        f_taylor_poly = f
+    else:
+        f_taylor_poly = f.series(x, n=m + n + 1).removeO().as_poly(x, field=True)
```
I think it would be better to require Expr rather than allowing Poly here as suggested by the type in the signature.

oscarbenjamin commented on sympy/series/approximants.py:
```diff
@@ -101,3 +108,282 @@ def approximants(l, X=Symbol('x'), simplify=False):
+    else:
+        f_taylor_poly = f.series(x, n=m + n + 1).removeO().as_poly(x, field=True)
+
+    """
+    TODO: if other methods for computing pade approximants are implemented,
+    this function should be modified to automatically select the best method.
+    """
```
Use `# comment` for comments.

oscarbenjamin commented on sympy/series/approximants.py:
```diff
@@ -101,3 +108,282 @@ def approximants(l, X=Symbol('x'), simplify=False):
+        yield r, t
+
+
+@public
+def pade_approximant_gcdex(
+    f: Poly, x: Expr, m: int, n: int | None = None
```
The argument `x` is redundant here if `f` is a Poly. Instead the function can just check that the `Poly` is univariate.

oscarbenjamin commented on sympy/series/approximants.py:
```diff
@@ -101,3 +108,282 @@ def approximants(l, X=Symbol('x'), simplify=False):
+    if m < 0 or n < 0:
+        raise ValueError("m and n must be non-negative integers")
+
+    min_degree = min(f.as_dict().keys())[0]
```
There is `f.EM()[0]` for this, not that it is implemented particularly efficiently.

pspoe commented on sympy/series/approximants.py:
```diff
@@ -101,3 +108,282 @@ def approximants(l, X=Symbol('x'), simplify=False):
+        raise ValueError("m and n must be non-negative integers")
+
+    min_degree = min(f.as_dict().keys())[0]
+    if min_degree > m:
```
what happens if you leave out this condition?
If  there is a solution where a power of x in nominator and denominatior can be cancelled I would call it acceptable if it meets the O(x**(n+m+1)) criterium

LylePoley replied:
This is intended to deal with some issues around polynomials of the form 

$$
P(x) = x^k\left(p_0 + p_1 x + p_2x^2 + \dots\right),
$$

where $p_0\neq 0$. In particular, there is no $[m/n]$ pade approximant for $P(x)$ if $m < k$.

pspoe replied:
I don't dispute your rationale.
But one lesson I learnt from `pade_approximant_lu` is that if you remove an up-front  condition the algorithm might either surprise you with an unexpected valid solution or take care in its own way (like raising a SingularMatrix exception).
If it is able to calculate a solution you can then examine it and raise an exception if you don't like it.

oscarbenjamin replied:
Yes, I wonder if it is better to return a result in this case. It seems Mathematica does. Are there other examples?

LylePoley replied:
As far as I can tell, Mathematica doesn't return results in this case. For example, the following 
```Wolfram
PadeApproximant[x(1 + x + x^2), {x, 0, {0, 1}}]
```
gives this error 
```
PadeApproximant::degnum: 2 The function x (1 + x + x ) has a zero of order 1 at 0 that is greater than the requested degree of the numerator.
```

LylePoley committed:
removed unnecessary function arguments

LylePoley committed:
fixing issues

LylePoley committed:
minor fixes

LylePoley committed:
minor fixes (2)

LylePoley committed:
minor fixes (3)

LylePoley commented:
I believe that all of the suggested changes have been implemented. When completed, the second approximant method from #27665 should now be easy to include.

LylePoley commented:
When the exception isn't included, the function will return None in those cases. I think that raising an exception which explains why a pade approximation fails is more informative.

LylePoley committed:
renamed `pade_approximants` to `pade_approximants_gcdex`

LylePoley commented:
I think this is ready.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -5307,6 +5314,120 @@ def half_gcdex(f, g, *gens, **args):
 
 
+@public
+def gcdex_steps(f, g):
```
I'm wondering if this should be made consistent with `gcdex` e.g.:
```python
In [3]: gcdex(x**2, x, x)
Out[3]: (0, 1, x)

In [4]: gcdex(Poly(x)**2, Poly(x))
Out[4]: (Poly(0, x, domain='QQ'), Poly(1, x, domain='QQ'), Poly(x, x, domain='QQ'))
```

LylePoley replied:
I think thats a good idea. However, I'm not entirely sure what is going on in the body of `gcdex` (sympy.polys.polytools line 5442)
```python
def gcdex(f, g, *gens, **args):
    options.allowed_flags(args, ['auto', 'polys'])

    try:
        (F, G), opt = parallel_poly_from_expr((f, g), *gens, **args)
    except PolificationFailed as exc:
        domain, (a, b) = construct_domain(exc.exprs)

        try:
            s, t, h = domain.gcdex(a, b)
        except NotImplementedError:
            raise ComputationFailed('gcdex', 2, exc)
        else:
            return domain.to_sympy(s), domain.to_sympy(t), domain.to_sympy(h)

    s, t, h = F.gcdex(G, auto=opt.auto)

    if not opt.polys:
        return s.as_expr(), t.as_expr(), h.as_expr()
    else:
        return s, t, h
```
If the 'Polyfication' succeeds, then I think I understand whats going on:
- First, the arguments `f` and `g` are converted to polynomials, and `opt` stores whether any conversion waas actially needed (i.e. if `f` and `g` were already of type `Poly`, then `opt.polys = True`). 
- Then, the member method `gcdex` is called to do the extended euclidean algorithm. 
- Finally, the function returns the result as `Poly` if the inputs were of type `Poly`, and as `Expr` if the inputs were not of type `Poly`.

However, I'm not sure what happens in the case where polyfication fails. I dont believe that `gcdex_steps` is implemented for any of the specific domains, so I assume that this would always run into a `NotImplementedError`? 

For consistency with `gcdex` and `half_gcdex`, should `gcdex_steps` be a member method of `Poly`?

oscarbenjamin replied:
When polification fails a routine for domain elements should be used. It does not have to be a method on every domain but just a function that works with domain elements rather than Poly. I'm not sure what domains it makes sense for but this is so that you can do e.g.:
```python
In [1]: gcdex(2, 4)
Out[1]: (1, 0, 2)

In [2]: gcdex(2*x, 4)
Out[2]: (0, 1/4, 1)
```
I would probably just make a function that assumes that it can use divmod with the domain elements and implements the extended Euclidean algorithm.

LylePoley replied:
When I run this with the current implementation (no fallback to working with domain elements) we get 
```python
for s in gcdex_steps(2*x, 4):
    print(s)
```
```
(0, 1, 4).
```
Is it problematic that the elements are not divided by `4`?

oscarbenjamin replied:
What domain does `gcdex_steps` use in this case? I think that `gcdex` always ensures that it is a polynomial with a field as ground domain.

LylePoley replied:
In this case, I believe that it will be a polynomial with the integers as the ground domain. As it is currently, the `gcdex_steps` attempts to convert the arguments to polynomials and throws an error if it can't, rather than falling back on an alternative. 

Why is it that `gcdex` ensures this before doing the calculation? Is it to guarantee that it is working with a euclidean domain?

oscarbenjamin replied:
> Is it to guarantee that it is working with a euclidean domain?

I assume so. Is the extended Euclidean algorithm defined for univariate polynomials with integer coefficients?

LylePoley replied:
$\mathbb{Z}[x]$ isn't a euclidean domain, so technically the answer is a no. [This stack exchange](https://math.stackexchange.com/questions/1300106/why-is-mathbbzx-not-a-euclidean-domain-what-goes-wrong-with-the-degree-f) link has some discussion about it.

I think the safest way is probably to just implement `gcdex_steps` in the same way as `gcdex`.

oscarbenjamin replied:
I agree.

oscarbenjamin replied:
It is the example I showed above: `gcdex(2, 4)`. Constructing polynomials fails if there are no symbols. Note:
```python
In [1]: gcdex(2, 4)
Out[1]: (1, 0, 2)

In [2]: gcdex(2, 4, x)
Out[2]: (0, 1/4, 1)
```
The first is gcdex in the domain `ZZ` and the latter is in `QQ[x]`. This one is `ZZ_I`:
```python
In [4]: gcdex(1+I, 2)
Out[4]: (-1, 1 + ⅈ, 1 + ⅈ)
```

LylePoley replied:
My mistake, I thought I had checked that case. So am I right in assuming that the `PolyficationFailed` except block is intended for dealing with the cases you have given? That is, when the arguments to `gcdex` aren't polynomials, but are elements of some domain?

I've re-structured `gcdex_steps` to be similar to `gcdex`. However, I haven't written any implementations of `gcdex_steps` for particular domains, so the `PolyficationFailed` branch will always raise `ComputationFailed` at the moment.

oscarbenjamin replied:
Yes, the PolificationFailed block is for when the expressions can be represented in a domain but are not polynomials. These interfaces are a little strange but that is how it works. Ideally polynomials would just be an example of a domain rather than something separate.

oscarbenjamin commented on sympy/series/approximants.py:
```diff
@@ -101,3 +108,283 @@ def approximants(l, X=Symbol('x'), simplify=False):
+    f: Poly, m: int, n: int | None = None
+    ) -> tuple[Poly, Poly] | tuple[None, None]:
+    """
+    `[m/n]` pade approximant of `f` around `x=0`. Computed using the extended euclidean algorithm.
+
+    Description
+    ===========
+
+    For a polynomial `f` and integers `m` and `n`, the `[m/n]` pade approximant
+    of `f` is a rational function `p(x)/q(x)`, where `p(x)` and `q(x)` are
+    polynomials of degree at most `m` and `n` respectively. The pade
+    approximation, when it exists, is such that the taylor series of `p(x)/q(x)`
+    around `x=0` matches `f` up to order `(m + n)` in `x`.
```
One of the changes I made when pushing a few commits here was to wrap long lines at 78 characters (https://github.com/sympy/sympy/pull/27671/commits/933b17f772f12f2abde38291d5287ce70807f260). This is particularly important to do in docstrings where there is no need to have such long lines and where the prose text is often displayed as tool tips in editors and things.

The first line here should be shortened rather than wrapped though.

oscarbenjamin commented on sympy/series/approximants.py:
```diff
@@ -101,3 +108,283 @@ def approximants(l, X=Symbol('x'), simplify=False):
+    return None, None
+
+
+@public
+def pade_approximant(
+    f: Expr, x: Expr, m: int, n: int | None = None
+    ) -> Expr | None:
+    """
+    `[m/n]` pade approximant of `f` around `x=0`.
```
This signature would make it awkward to add an `x0` parameter in future. Maybe that parameter should be added now like for Mathematica's PadeApproximant function.

LylePoley committed:
Refactoring pade approximants

1. Added the option of expanding around `x=x0`.
2. `gcdex_steps` can take and return `Poly` or `Expr`, consistent with th existing`gcdex` functionality.
3. Edited tests/doctests to reflect changes.

LylePoley committed:
Removed type hints

LylePoley committed:
Minor fixes

LylePoley committed:
Docstring fixes

LylePoley committed:
Docstring fixes

LylePoley committed:
Added x0 to `pade_approximant` docstring

LylePoley committed:
Aligned structure of `gcdex_steps` with `gcdex`

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -5307,6 +5315,130 @@ def half_gcdex(f, g, *gens, **args):
+    except PolificationFailed as exc:
+        domain, (a, b) = construct_domain(exc.exprs)
+
+        try:
+            steps = domain.gcdex_steps(a, b)
+        except NotImplementedError:
+            raise ComputationFailed('gcdex_steps', 2, exc)
+        else:
```
I would just make a private function for working with domain elements rather than calling a method on the domain:
```python
def _gcdex_steps_domain(a, b, K):
    if not K.is_PID:
        raise DomainError("gcdex is only for Euclidean domains")
    # implementation here
```
With domains you can check for zero with just `if e:` and you can do Euclidean division with:
```python
In [11]: ZZ.div(7, 2)
Out[11]: (3, 1)
```

LylePoley replied:
OK that makes sense. Correct me if I'm wrong, but I think that there are PID's that are not euclidean domains?

Also, should this function be put into `polys.polytools` or somewhere else?

oscarbenjamin replied:
There is no way to check for a Euclidean domain. I'm not sure what would be an example of a PID that is not a Euclidean domain but I don't think that there are any in SymPy.

If it is a private function I would just put it in polytools next to the `gcdex_steps` function since that is the only place it would be used.

LylePoley replied:
OK, I'll do that. 

Is there any difference between something like `ZZ.div(7, 2)` and `divmod(7, 2)`? If not, should either be preferred?

oscarbenjamin replied:
The domain methods should always be used for division when working with domain elements. Here is one example of a difference:
```python
In [18]: EX.div(EX(x**2-1), EX(x-1))
Out[18]: (EX(x + 1), EX(0))

In [19]: divmod(EX(x**2-1), EX(x-1))
---------------------------------------------------------------------------
TypeError
```

LylePoley committed:
Added `_gcdex_steps_domain` helper function.

oscarbenjamin commented on sympy/polys/tests/test_polytools.py:
```diff
@@ -1928,6 +1929,18 @@ def test_issue_7864():
     assert r == 0
 
 
+def test_gcdex_steps():
+    f = Poly(x**5 + 2*x**4 - x**2 + 1, x)
+    g = Poly(x**4 - 1, x)
+    eea_result = gcdex_steps(f, g)
+
+    r_degree = 5
+    for si, ti, ri in eea_result:
+        assert si*f + ti*g == ri
+        assert ri.degree() < r_degree
+        r_degree = ri.degree()
```
I think this needs a few more tests. At least there should be a test with Expr inputs and a test that uses the domain implementation.

oscarbenjamin commented:
Apart from the comment above about tests I think this looks good. Is there anything left outstanding?

LylePoley commented:
The only thing I can think of is in the following example. First, in the case where the polyfication fails, both `gcdex` and `gcdex_steps` fall back on the algorithm applied to a domain and get the same result
```python
gcdex(2, 5)
```
```
(-2, 1, 1)
```
```python
for step in gcdex_steps(2, 5):
    print(step)
```
```
(0, 1, 5)
(1, 0, 2)
(-2, 1, 1).
```
However, when the arguments _can_ be turned into polynomials, the results are only the same up to a constant
```python
gcdex(2*x, 5)
```
```
(0, 1/5, 1)
```
```python
for step in gcdex_steps(2*x, 5):
    print(step)
```
```
(0, 1, 5).
```
Both solutions are correct, as the `gcd` of two polynomials is only unique up to multiplication by an element of the domain. Is there any rationale to making them the same by dividing by the leading coefficient of the gcd at the end in `gcdex_steps`?

oscarbenjamin commented:
> Is there any rationale to making them the same by dividing by the leading coefficient of the gcd at the end in `gcdex_steps`?

I'm not sure. It seems that gcd is generally defined here as being monic when over a field:
```python
In [1]: gcdex(2*(x + 2), 4*(x + 2))
Out[1]: (0, 1/4, x + 2)

In [2]: gcd(2*(x + 2), 4*(x + 2))
Out[2]: 2⋅x + 4

In [3]: gcd(2*(x + 2), 4*(x + 2), domain=QQ)
Out[3]: x + 2
```
It is often a bad idea to make it monic though:
```python
In [7]: gcd(2*((y+1)*x + 2), 4*((y+1)*x + 2), x, field=True)
Out[7]: 
      2  
x + ─────
    y + 1

In [8]: gcd(2*((y+1)*x + 2), 4*((y+1)*x + 2), x)
Out[8]: x⋅(2⋅y + 2) + 4
```
Generally I prefer to have a primitive denominator free polynomial rather than a monic one.

It looks like it is just made monic as an extra step at the end rather than a consequence of the algorithm:
https://github.com/sympy/sympy/blob/98f811f58012a904b34299d988f46c836cb79f89/sympy/polys/euclidtools.py#L74-L82
I would probably have preferred not to do that but I'm not sure how consistent these things are across the many polynomial types and domains.

oscarbenjamin commented:
I think that it is fine that `gcdex_steps` does not make the output monic. It also seems more consistent e.g.:
```python
In [11]: gcdex(2, 4)
Out[11]: (1, 0, 2)

In [12]: gcdex(2*x, 4*x)
Out[12]: (0, 1/4, x)

In [13]: list(gcdex_steps(2, 4))
Out[13]: [(0, 1, 4), (1, 0, 2)]

In [14]: list(gcdex_steps(2*x, 4*x))
Out[14]: [(0, 1, 4⋅x)]
```
That last one is also consistent with gcd of a polynomial over a field being defined only up to multiplication by a unit. Presumably making it monic would not improve the formulae given for the Pade approximants.

LylePoley commented:
The multivatiate polynomial is a good example. I agree with you that there is no need to make the gcd monic.

I've also ran into some possibly unexpected behaviour in `gcdex`, which `gcdex_steps` mimics
```ptyhon
gcdex(sqrt(2)*x, sqrt(2) + x)
```
```
    raise ValueError('univariate polynomial expected')
ValueError: univariate polynomial expected
```
The reason this happens is because `sqrt(2)*x` is polyfied to 
```python 
Poly(sqrt(2)*x, sqrt(2), x, 'ZZ')
```
rather than something like 
```python
Poly(sqrt(2)*x, x, ZZ[sqrt(2)])
```

I'm not suggesting that anything be done about it in this PR, but thought I'd mention it.

oscarbenjamin commented:
You would need to specify the generator or pass extension=True:
```python
In [3]: gcdex(sqrt(2)*x, sqrt(2) + x, x)
Out[3]:
⎛      √2   ⎞
⎜-1/2, ──, 1⎟
⎝      2    ⎠

In [4]: gcdex(sqrt(2)*x, sqrt(2) + x, extension=True)
Out[4]:
⎛      √2   ⎞
⎜-1/2, ──, 1⎟
⎝      2    ⎠
```
There are a bunch of options that control these things:
```python
In [8]: Poly(sqrt(2)*x)
Out[8]: Poly(x*(sqrt(2)), x, sqrt(2), domain='ZZ')

In [9]: Poly(sqrt(2)*x, greedy=False)
Out[9]: Poly(sqrt(2)*x, x, domain='EX')

In [10]: Poly(sqrt(2)*x, x)
Out[10]: Poly(sqrt(2)*x, x, domain='EX')

In [11]: Poly(sqrt(2)*x, extension=True)
Out[11]: Poly(sqrt(2)*x, x, domain='QQ<sqrt(2)>')
```
The greedy behaviour is useful in various places. Ideally we would just use `extension=True` by default in places like this but it can be slow sometimes to construct the extension field.

The `pade_approximants` etc functions should specify the generator when constructing Poly or when calling `gcdex_steps` etc.

LylePoley committed:
adding to `gcdex_steps` and its tests

oscarbenjamin commented:
I'm not sure what you mean. There should be one place that converts the Expr into Poly and at that place the generator should be specified. In the other functions the expectation should just be that the inputs are univariate Poly with the given generator.

LylePoley commented:
As it is, `pade_approximant_gcdex` and `pade_approximants_gcdex` can have `Poly` or `Expr` as arguments. If given arguments of type `Expr` they convert them to `Poly` before passing them to `gcdex_steps`. I misunderstood your [comment](https://github.com/sympy/sympy/pull/27671#discussion_r2087376794) about making the behaviour of `gcdex_steps` consistent with `gcdex` and did the same with the functions `pade_approximant_gcdex` and `pade_approximants_gcdex` as well. I'll reverse that.

For clarity, here are the the signatures of the functions currently
```python
gcdex_steps(f: Poly | Expr, g: Poly | Expr, *gens, **args) -> Generator[tuple[Poly, Poly, Poly]] | Generator[tuple[Expr, Expr, Expr]]
pade_approximants_gcdex(f: Poly | Expr, order: int) -> Generator[tuple[Poly, Poly]] | Generator[tuple[Expr, Expr]]
pade_approximant_gcdex(f: Poly | Expr, n: int, m: int) -> tuple[Poly, Poly] | tuple[Expr, Expr]
pade_approximant(f: Poly | Expr, x, x0, n, m) -> Expr
```
and here is how they will look after the changes
```python
gcdex_steps(f: Poly | Expr, g: Poly | Expr, *gens, **args) -> Generator[tuple[Poly, Poly, Poly]] | Generator[tuple[Expr, Expr, Expr]]
pade_approximants_gcdex(f: Poly, order: int) -> Generator[tuple[Poly, Poly]]
pade_approximant_gcdex(f: Poly, n: int, m: int) -> tuple[Poly, Poly]
pade_approximant(f: Poly | Expr, x, x0, n, m) -> Expr
```

oscarbenjamin commented:
I'm not too keen on the complex type signatures for functions like `gcdex` but I think it makes sense that `gcdex_steps` matches them.

For `pade_approximants` I would say that one function should be denoted the "high-level" interface and that one should take Expr and a specified generator. The conversion from Expr to Poly should be done in one place and then all the other functions should just take univariate Poly and not do any conversions.

LylePoley commented:
In the above, `pade_approximant` is the high level function, so that will handle conversion to `Poly`. I'll remove any conversion code from the `..._gcdex` functions. Or are you suggesting another function `pade_approximants` as an interface for `pade_approximants_gcdex` which can handle conversion between `Expr` and `Poly`?

oscarbenjamin commented:
> Or are you suggesting another function

No, I just forgot what the function was called.

LylePoley committed:
Changed `..._gcdex` pade functions to accept Polys rather than Expr.

LylePoley committed:
Minor fixes

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -5307,6 +5315,182 @@ def half_gcdex(f, g, *gens, **args):
+        t1, t2 = t2, t1 - quotient * t2
+
+
+def _gcdex_steps_polynomial(f, g, auto):
+    """
+    Generator for intermediate steps of the extended euclidean algorithm
+    on polynomials. Helper function for `gcdex_steps`.
+    """
+    if auto and f.get_domain().is_Ring:
+        f, g = f.to_field(), g.to_field()
+
+    if not f.get_domain().is_PID:
+        raise DomainError("gcdex_steps is only for Euclidean domains")
+
+    if not f.is_univariate:
+        raise ValueError('univariate polynomial expected')
```
I think that the `is_PID` check here should really be `if f.domain.is_Field`. Univariate polynomials over a field are suitable for `gcdex`.

LylePoley committed:
changed `.is_PID` to `.is_Field` in `_gcdex_steps_polynomial`

oscarbenjamin commented on sympy/polys/domains/domain.py:
```diff
@@ -1287,6 +1287,10 @@ def half_gcdex(self, a, b):
         s, t, h = self.gcdex(a, b)
         return s, h
 
+    def gcdex_steps(self, a, b):
+        """Intermediate steps of the extended GCD algorithm. """
+        raise NotImplementedError
```
This should be removed because it is a function now rather than a domain method.

oscarbenjamin commented:
I think this looks good now apart from the last comment. It is important that we don't add a domain method if we do not intend to implement it for all the domains.

LylePoley committed:
Removed `gcdex_steps` from `Domain`

oscarbenjamin commented on sympy/series/tests/test_approximants.py:
```diff
@@ -21,3 +23,96 @@ def test_approximants():
+        (x-1)*(11*(x-1)**2 + 60*(x-1) + 60)/(3*((x-1)**3 + 12*(x-1)**2 + 30*(x-1) + 20))
+        ).simplify() == 0
+
+    is_error = False
+    try:
+        pade_approximant(sin(x), x, 0, 0, 4)
+    except ValueError:
+        is_error = True
+    assert is_error
```
Sorry I missed this in previous passes. This should use raises:
```
from sympy.testing.pytest import raises

with raises(ValueError):
     pade_approximant(sin(x), x, 0, 0, 4)
```

LylePoley replied:
I used the syntax 
```python 
raises(ValueError, lambda : pade_approximant(sin(x), x, 0, 0, 4))
```
as pylance doesn't seem to like 
```python
with raises(ValueError):
     pade_approximant(sin(x), x, 0, 0, 4)
```
It says this 
```
Object of type "ExceptionInfo" cannot be used with "with" because it does not correctly implement __enter__
  Attribute "__enter__" is unknownPylance[reportGeneralTypeIssues](https://github.com/microsoft/pylance-release/blob/main/docs/diagnostics/reportGeneralTypeIssues.md)
```

oscarbenjamin replied:
I've fixed the type annotations for raises and also added annotations for all the new functions here.

LylePoley committed:
Changed error test to use `raises`

oscarbenjamin committed:
typ: add type annotations for raises and gcdex_steps

oscarbenjamin committed:
doc: fix docstring for gcdex_steps

oscarbenjamin commented on sympy/series/approximants.py:
```diff
@@ -101,3 +110,292 @@ def approximants(l, X=Symbol('x'), simplify=False):
+    =======
+
+    pade approximant : tuple[Poly, Poly]
+        returns p(x)
```
This is not true for `pade_approximant` as it returns `Expr | None`.

I think it would be good to explain an example of the None case in the docstring.

LylePoley replied:
I think that with the changes that have been made over the last however many commits, `Pade_approximant` will never return `None`. The cases which used to return `None` now raise errors with a description of why.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -5307,6 +5316,202 @@ def half_gcdex(f, g, *gens, **args):
+    =======
+
+    generator : Generator[tuple[Poly, Poly, Poly]] | Generator[tuple[Expr, Expr, Expr]]
+        A generator to the sequences `s`, `t`, and `r`
```
It is probably better to say iterator rather than generator.

oscarbenjamin commented:
The release notes in the OP should be updated and expanded a little to explain the `pade_approximant` function which is the main end user function here.

LylePoley committed:
Remove `None` return from `pade_approximant`

LylePoley committed:
Minor fixes

LylePoley commented:
To change the release notes, do I just edit my original comment?

oscarbenjamin commented on sympy/series/approximants.py:
```diff
@@ -101,3 +110,289 @@ def approximants(l, X=Symbol('x'), simplify=False):
+    if m < 0 or n < 0:
+        raise ValueError(f"{m=} and {n=} must be non-negative integers")
+
+    f_taylor_poly = f.series(x, x0, n=m + n + 1).subs(x, x + x0)\
+                    .removeO().as_poly(x, field=True)
```
I think that the previous code here was better:
```python
f.subs(x, x + x0).series(x, 0, ...).removeO().as_poly(x)
```
Internally if `x0 != 0` then series just does the same substitution and back so this ends up doing three substitutions rather than one.

oscarbenjamin commented:
> To change the release notes, do I just edit my original comment?

Yes

LylePoley committed:
Expanded `pade_approximant` to deal with functions with laurent expansions.

LylePoley committed:
Minor fixes

LylePoley commented:
The examples I have tested agree with mathematica.

oscarbenjamin commented:
I just looked through and it looks good to me. Are you done with it now?

oscarbenjamin commented:
> Would you like me to elaborate on the use of `pade_approximant` some more?

It should have some explanation of what the functions are and what feature they provide.

Many of sympy's release notes are not well written but you can see the notes for 1.14 here:
https://github.com/sympy/sympy/wiki/Release-Notes-for-1.14

Here is a somewhat comparable example:

> A new function factor_system() is added to factorize a system of equations into subsystems consisting of only irreducible factors. Functions factor_system_bool(), factor_system_cond and factor_system_poly are also added to express a factorized system of equations in disjunctive normal form while preserving all information about degenerate cases of any symbolic parameters.

LylePoley commented:
No problem I've enjoyed it! I'm interested in getting involved in improving the functionality of `series`, so you may be seeing more of me.

Edited the release notes, let me know if anything there needs more clarification.

oscarbenjamin commented:
I removed the LaTeX markers from the release notes because I'm not sure they will work in other places the release notes are displayed. Otherwise looks good though.

Okay, let's get this in. Thanks!
